### ISSUE TITLE
Precision Loss in ConfusionMatrix

### ISSUE THREAD

**Issue by zhengzangw** [PyTorchLightning/metrics#187]
## 🐛 Bug

In Pytorch, `torch.zeros` and `.float()` methods create or convert a tensor to be a `torch.float32` tensor. Compare to the default `np.float64` data type in NumPy, it cannot provide enough precision. In `ConfusionMatrix`, the default values are torch.float32, which leads to wrong calculation (although maybe very little) against ground truth.

I think we only need to change `torch.zeros(num_classes, num_classes)` to `torch.zeros(num_classes, num_classes, dtype=torch.float64)`. I believe this change introduces negligible memory cost.

### To Reproduce

```python
>>> torch.tensor(90170814.).item()
90170816.0
```

### Expected Behaviour

```python
>>> torch.tensor(90170814., dtype=torch.float64).item()
90170814.0
```

### Additional context

`ConfusionMatrix` is the base of many other metrics, such as `IoU`. I found this problem by comparing the results to another NumPy-based metric in a semantic segmentation task using `IoU` as the metric. I think in today's deep learning dataset, float32 precision is not enough for the implementation of metrics.

Based on the above observation, maybe all `PyTorchLightning/metrics` should be implemented based on `torch.float64`

**Comment by zhengzangw** [comment]
Considering we may not need an accurate metric when validating (I see that even 16-precision is supported), maybe a flag to enable 64-precision in testing is a better choice.

**Comment by SkafteNicki** [comment]
Hi @zhengzangw,
So in general we want to stick with the defaults of pytorch. In general the field seems to be moving towards using tensors with dtype 16 to save memory. That said, most of our metric should be dtype agnostic meaning that you can just cast the metric to whatever dtype that you want to use:
```python
c = ConfusionMatrix(num_classes=3).double()
c(preds, target) # output will be double precision
``` 
After playing around with this, I identified a bug that currently prevents this but should be fixed by PR 243. Thanks for opening this issue such that we found this bug :]

**Comment by zhengzangw** [comment]
I see. Since there is a way to maintain double precision, I will close the issue.
